src/Middleware/HttpLogging/src/RequestBufferingStream.cs

```diff
@@ -27,6 +27,12 @@ public override async ValueTask<int> ReadAsync(Memory<byte> destination, Cancell
     {
         var res = await _innerStream.ReadAsync(destination, cancellationToken);
 
+        // Zero-byte reads can occur when using PipeReader, we don't want to accidentally complete the RequestBody logging in this case
+        if (destination.IsEmpty)
```

Comment:
What about the one on [line 60](https://github.com/dotnet/aspnetcore/pull/47794/files#diff-9d3086eac5e1da5350c127d51efbb6f16fa3134e4337fbed15e3a07bb8f2b940R60)?

Reply:
If we checked `res` instead we'd still need to check if `destination.IsEmpty`, because you could reach the end of the underlying stream which would give `res == 0`. So it's simpler just to check `destination.IsEmpty`.

Reply:
I guess the one on line 60 should be updated for consistency. The reason I originally didn't change it is because zero-byte reads are almost always done via PipeReader, which doesn't have a synchronous read so Read won't be called. And, by default ASP.NET Core doesn't allow synchronous reads on streams anyways.

But, if a user set the option to allow synchronous reads, and did `request.Body.Read(new byte[0])`, then they would hit this path. It would be really odd, since the whole point of zero-byte reads is to avoid allocating until the underlying stream has data, which would be an async operation if there was no data yet.

TLDR; I'll update it for consistency 😄

Reply:
I was not reading this correctly.  "zero-byte reads" refers to the case where the destination buffer is empty, rather than to the case where the stream writes zero bytes to the buffer.

Reply:
Zero-byte reads allow code to wait for available data without providing a potentially big buffer while waiting.

For example:
1. Make zero byte read to wait for data to be available: `await stream.ReadAsync(Array.Empy<byte>())`
2. Zero-byte read returns. Rent buffer from pool and read again with a buffer: `await stream.ReadAsync(oneMegabyteBuffer)`. It should return quickly, allowing the buffer to be returned to the pool quickly.

Reply:
> TLDR; I'll update it for consistency 😄

Are zero-byte reads for sync `Stream.Read` a thing? I haven't use sync read methods in so long that I'm not sure.

Reply:
I'm fine with asserting it never happens if that's the case.

Reply:
> Are zero-byte reads for sync Stream.Read a thing?

I think they are, because lots of stream implementations just call the async api and block, so they inherit the zero-byte read feature from the async code path.

Reply:
Yes, be consistent for Read and ReadAsync.